### PR TITLE
[SPARK-32521][SQL] Bug-fix: WithFields Expression should not be foldable

### PR DESCRIPTION
**Pull request by fqaiser94**
### What changes were proposed in this pull request?

Make WithFields Expression not foldable. 

### Why are the changes needed?

The following query currently fails on master brach: 
```
sql("SELECT named_struct('a', 1, 'b', 2) a")
.select($"a".withField("c", lit(3)).as("a"))
.show(false)
// java.lang.UnsupportedOperationException: Cannot evaluate expression: with_fields(named_struct(a, 1, b, 2), c, 3)
```
This happens because the Catalyst optimizer sees that the WithFields Expression is foldable and tries to statically evaluate the WithFields Expression (via the ConstantFolding rule), however it cannot do so because WithFields Expression is Unevaluable. 

### Does this PR introduce _any_ user-facing change?

Yes, queries like the one shared above will now succeed. 
That said, this bug was introduced in Spark 3.1.0 which has yet to be released. 

### How was this patch tested?

A new unit test was added.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 1]
```diff
@@ -1420,4 +1420,16 @@ class ColumnExpressionSuite extends QueryTest with SharedSparkSession {
       }.getMessage should include("No such struct field b in a, B")
     }
   }
+
+  test("SPARK-32521: withField with statically evaluable arguments should succeed") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:1424]: The end users can see the examples. To ensure we do not have a regression, could you add all the examples to the test suite?

Comment (fqaiser94) [reply sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:1456]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 2]
```diff
@@ -1420,4 +1420,48 @@ class ColumnExpressionSuite extends QueryTest with SharedSparkSession {
...
+        .select($"struct_col".withField("a.c", lit(3))),
+      Row(Row(Row(1, 2, 3))))
+
+    intercept[AnalysisException] {
+      sql("SELECT named_struct('a', named_struct('b', 1), 'a', named_struct('c', 2)) struct_col")
+        .select($"struct_col".withField("a.c", lit(3)))
+    }.getMessage should include("Ambiguous reference to fields")
+  }
+
+  test("SPARK-32521: withField with statically evaluable arguments should succeed") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:1456]: Do we still need this test case? It's included in `withField user-facing examples`.

Comment (fqaiser94) [reply sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: You're right, removed.

**Comment by cloud-fan** [comment]
the last commit just removes one test, we don't need to wait for jenkins.

Thanks, merging to master!

**Pull request closed without merging**